File: src/main/java/com/arnav/tutionSAAS/controller/AttendanceController.java
package com.arnav.tutionSAAS.controller;

import com.arnav.tutionSAAS.dto.AttendanceRequest;
import com.arnav.tutionSAAS.dto.AttendanceResponse;
import com.arnav.tutionSAAS.service.AttendanceService;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.ResponseEntity;
import org.springframework.security.access.prepost.PreAuthorize;
import org.springframework.security.core.annotation.AuthenticationPrincipal;
import org.springframework.security.oauth2.jwt.Jwt;
import org.springframework.web.bind.annotation.*;

@RestController
public class AttendanceController {

    @Autowired private AttendanceService attendanceService;

    @PostMapping("/api/sessions/{sessionId}/attendance")
    @PreAuthorize("hasRole('TEACHER')")
    public ResponseEntity<AttendanceResponse> markAttendance(
            @PathVariable Long sessionId,
            @RequestBody AttendanceRequest request,
            @AuthenticationPrincipal Jwt jwt) {
        return ResponseEntity.ok(attendanceService.markAttendance(sessionId, request, jwt.getSubject()));
    }

    @GetMapping("/api/sessions/{sessionId}/attendance")
    @PreAuthorize("hasAnyRole('TEACHER', 'ADMIN')")
    public ResponseEntity<AttendanceResponse> getAttendance(@PathVariable Long sessionId) {
        return ResponseEntity.ok(attendanceService.getAttendanceForSession(sessionId));
    }

    @GetMapping("/api/batches/{batchId}/students/{studentId}/attendance")
    @PreAuthorize("hasAnyRole('TEACHER', 'ADMIN')")
    public ResponseEntity<AttendanceService.AttendanceSummary> getStudentSummary(
            @PathVariable Long batchId,
            @PathVariable Long studentId) {
        return ResponseEntity.ok(attendanceService.getStudentAttendanceSummary(batchId, studentId));
    }
}

File: backend-nest/src/attendance/attendance.controller.ts
import {
  Body,
  Controller,
  Get,
  Param,
  ParseIntPipe,
  Post,
  UseGuards,
} from '@nestjs/common';
import { Role } from '@prisma/client';
import { ClerkAuthGuard } from '../common/guards/clerk-auth.guard';
import { RolesGuard } from '../common/guards/roles.guard';
import { Roles } from '../common/decorators/roles.decorator';
import { ClerkId } from '../common/decorators/current-user.decorator';
import { AttendanceService } from './attendance.service';

@Controller()
@UseGuards(ClerkAuthGuard, RolesGuard)
export class AttendanceController {
  constructor(private readonly attendanceService: AttendanceService) {}

  @Post('api/sessions/:sessionId/attendance')
  @Roles(Role.TEACHER)
  markAttendance(
    @Param('sessionId', ParseIntPipe) sessionId: number,
    @Body() body: any,
    @ClerkId() clerkId: string,
  ) {
    return this.attendanceService.markAttendance(sessionId, body, clerkId);
  }

  @Get('api/sessions/:sessionId/attendance')
  @Roles(Role.TEACHER, Role.ADMIN)
  getAttendance(@Param('sessionId', ParseIntPipe) sessionId: number) {
    return this.attendanceService.getAttendanceForSession(sessionId);
  }

  @Get('api/batches/:batchId/students/:studentId/attendance')
  @Roles(Role.TEACHER, Role.ADMIN, Role.STUDENT)
  getStudentSummary(
    @Param('batchId', ParseIntPipe) batchId: number,
    @Param('studentId', ParseIntPipe) studentId: number,
  ) {
    return this.attendanceService.getStudentAttendanceSummary(batchId, studentId);
  }
}
